File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.py
def solution(bandage, health, attacks):
    continuity = 0 # 연속 성공
    last_attack_time = max(attacks)[0] # 마지막 공격이 이루어지는 시간초
    attack_dict = {}
    max_health = health
    
    for i in attacks:
        attack_dict[i[0]] = i[1]
    
    for i in range(1, last_attack_time + 1):
        if i in attack_dict:
            health -= attack_dict[i]
            continuity = 0
            if health <= 0:
                return -1
        else:
            continuity += 1
            if continuity < bandage[0]:
                health += bandage[1]
                if health > max_health:
                    health = max_health
            else:
                health += bandage[1] + bandage[2]
                continuity = 0
                if health > max_health:
                    health = max_health

    return health

File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.java
import java.util.*;

class Solution {
    public int solution(int[] bandage, int health, int[][] attacks) {
        int continuity = 0;
        int lastAttackTime = 0;
        Map<Integer, Integer> attackMap = new HashMap<>();
        int maxHealth = health;
        
        for (int[] attack : attacks) {
            lastAttackTime = Math.max(lastAttackTime, attack[0]);
            attackMap.put(attack[0], attack[1]);
        }
        
        for (int i = 1; i <= lastAttackTime; i++) {
            if (attackMap.containsKey(i)) {
                health -= attackMap.get(i);
                continuity = 0;
                if (health <= 0) {
                    return -1;
                }
            } else {
                continuity++;
                if (continuity < bandage[0]) {
                    health += bandage[1];
                    if (health > maxHealth) {
                        health = maxHealth;
                    }
                } else {
                    health += bandage[1] + bandage[2];
                    continuity = 0;
                    if (health > maxHealth) {
                        health = maxHealth;
                    } 
                }
            }
        }
        
        return health;
    }
}